src/main/java/bio/terra/service/filedata/flight/ingest/IngestBulkGcpStep.java

```diff
@@ -0,0 +1,520 @@
+    this.fileDao = fileDao;
+    this.fileService = fileService;
+    this.executor = executor;
+    this.maxPerformanceThreadQueueSize = maxPerformanceThreadQueueSize;
+  }
+
+  /**
+   * Return a stream of file models representing the files to ingest. The construction of this
+   * stream varies based on the type of ingestion flight
+   */
+  protected abstract Stream<BulkLoadFileModel> getModelsStream(FlightContext context);
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+
+    initializeResultsModel(context.getFlightId(), loadTag);
+
+    FlightMap workingMap = context.getWorkingMap();
+
+    GoogleBucketResource bucketResource =
+        FlightUtils.getContextValue(context, FileMapKeys.BUCKET_INFO, GoogleBucketResource.class);
+    List<FileModel> ingestedFiles;
+    try (var loadModels = getModelsStream(context)) {
+      ingestedFiles = performIngest(workingMap, loadModels.toList(), bucketResource);
+    } catch (InvalidUserProjectException ex) {
+      // We retry this exception because often when we've seen this error it has been transient
+      // and untruthful -- i.e. the user project specified exists and has a legal id.
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, ex);
+    } catch (GoogleInternalServerErrorException ex) {
+      // Google's error message suggests retrying the operation
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, ex);
+    } catch (FileSystemAbortTransactionException rex) {
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, rex);
+    }
+
+    logger.info("Recording results");
+    // Must convert to set so that stairway can deserialize
+    workingMap.put(IngestMapKeys.COMBINED_EXISTING_FILES, new HashSet<>(ingestedFiles));
+    workingMap.put(
+        IngestMapKeys.BULK_LOAD_RESULT,
+        new BulkLoadArrayResultModel().loadSummary(resultModel).loadFileResults(loadedFiles));
+
+    logger.info("Done ingesting files");
+    return StepResult.getStepResultSuccess();
+  }
+
+  /** Main driving method that performs the file ingestion */
+  private List<FileModel> performIngest(
+      FlightMap workingMap, List<BulkLoadFileModel> loadModels, GoogleBucketResource bucketResource)
+      throws InterruptedException {
+
+    resultModel.totalFiles(loadModels.size());
+    List<BulkLoadFileModel> successfulLoadModels;
+    List<FSFileInfo> fsFileInfos = new ArrayList<>();
+    Map<String, UUID> fileIdsByPath = new HashMap<>();
+
+    // In the case where file ids are predictable, the linking / copying of the data must happen
+    // first in order to obtain file ids
+    if (dataset.hasPredictableFileIds()) {
+      List<CopyResult> copyResults =
+          doFileCopy(workingMap, loadModels, fileIdsByPath, bucketResource);
+
+      fsFileInfos.addAll(
+          copyResults.stream()
+              .filter(r -> Objects.isNull(r.error()) && Objects.nonNull(r.fsFileInfo()))
```

Comment:
The `isNull` and `nonNull` methods are intended to be used as functional arguments If you want to compare against `null` directly, `!= null` is better IMO.
```suggestion
              .filter(r -> r.error() == null && r.fsFileInfo() != null)
```